engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/RegionedColumnSourceManager.java

```diff
@@ -834,34 +833,46 @@ public void pushdownFilter(
-                                if (result.match().isNonempty()) {
-                                    synchronized (matchBuilder) {
-                                        result.match().shiftInPlace(locationStartKey);
-                                        matchBuilder.addRowSet(result.match());
-                                    }
+                                // TODO: This is a bit dangerous, but seems "ok" in context?
+                                if (match.isNonempty()) {
+                                    match.shiftInPlace(locationStartKey);
                                 }
-                                if (result.maybeMatch().isNonempty()) {
-                                    synchronized (maybeMatchBuilder) {
-                                        result.maybeMatch().shiftInPlace(locationStartKey);
-                                        maybeMatchBuilder.addRowSet(result.maybeMatch());
-                                        maybeMatchCount.add(result.maybeMatch().size());
-                                    }
+                                if (maybeMatch.isNonempty()) {
+                                    maybeMatch.shiftInPlace(locationStartKey);
                                 }
+                                builder.add(locationIdx, locationInputSize, result);
                             }
                         };
+                        // TODO: fullSet is not shifted here, and is a lie?
+                        // Should it actually be a subset in the parallelized case?
+                        // Or, can we get rid of it?
+                        // fullSet.subSetByKeyRange(locationStartKey, locationEndKey).shiftInPlace(-locationStartKey)?
                         entry.location.pushdownFilter(filter, renameMap, locationInput, fullSet, usePrev, context,
                                 costCeiling, jobScheduler, resultConsumer, onError);
                     }
                     locationResume.run();
-                }, () -> onComplete.accept(PushdownResult.of(matchBuilder.build(),
-                        maybeMatchCount.get() == input.size() ? input.copy() : maybeMatchBuilder.build())),
+                }, () -> {
+                    try (
+                            final RowSet ignore = selectionCopy;
+                            final Builder ignore2 = builder) {
+                        onComplete.accept(builder.build(selectionCopy));
+                    }
+                },
+                // Note: we should technically be closing in the error case as well; but in the error case, we may
+                // be less concerned, and happy enough for GC to take care of it. I worry a bit about the contract
+                // around onError, as it looks like we are giving each parallel task a chance to call it, so it
+                // doesn't look like we are guarding it to only call it once (and double-closing a RowSet is much
```

Comment:
Error handling in the job scheduler is pretty sophisticated. The `onError` call will be invoked only once after currently executing tasks are complete. So this error will propagate only once.